cores/esp8266/core_esp8266_waveform.c

```diff
@@ -149,122 +104,153 @@ void setTimer1Callback(uint32_t (*fn)()) {
+// Normally would not want two copies like this, but due to different
+// optimization levels the inline attribute gets lost if we try the
+// other version.
+
+static inline ICACHE_RAM_ATTR uint32_t GetCycleCountIRQ() {
+  uint32_t ccount;
+  __asm__ __volatile__("rsr %0,ccount":"=a"(ccount));
+  return ccount;
+}
+
+static inline ICACHE_RAM_ATTR uint32_t min_u32(uint32_t a, uint32_t b) {
+  if (a < b) {
+    return a;
+  }
+  return b;
+}
+
 // Stops a waveform on a pin
 int ICACHE_RAM_ATTR stopWaveform(uint8_t pin) {
   // Can't possibly need to stop anything if there is no timer active
   if (!timerRunning) {
     return false;
   }
-
-  for (size_t i = 0; i < countof(waveform); i++) {
-    if (!waveform[i].enabled) {
-      continue; // Skip fast to next one, can't need to stop this one since it's not running
-    }
-    if (((pin == 16) && waveform[i].gpio16Mask) || ((pin != 16) && (waveform[i].gpioMask == 1<<pin))) {
-      // Note that there is no interrupt unsafety here.  The IRQ can only ever change .enabled from 1->0
-      // We're also doing that, so even if an IRQ occurred it would still stay as 0.
-      waveform[i].enabled = 0;
-      int cnt = timer1CB ? 1 : 0;
-      for (size_t i = 0; (cnt == 0) && (i < countof(waveform)); i++) {
-        cnt += waveform[i].enabled ? 1 : 0;
-      }
-      if (!cnt) {
-        deinitTimer();
-      }
-      return true;
-    }
+  // If user sends in a pin >16 but <32, this will always point to a 0 bit
+  // If they send >=32, then the shift will result in 0 and it will also return false
+  uint32_t mask = 1<<pin;
+  if (!(waveformEnabled & mask)) {
+    return false; //It's not running, nothing to do here
   }
-  return false;
+  waveformToDisable |= mask;
+  // Ensure tiomely service....
+  if (T1L > MicrosecondsToCycles(10)) {
+    timer1_write(MicrosecondsToCycles(10));
+  }
+  while (waveformToDisable) {
+    delay(1); // Wait for IRQ to update
+  }
+  if (!waveformEnabled && !timer1CB) {
+    deinitTimer();
+  }
+  return true;
 }
 
+#if F_CPU == 80000000
+  #define DELTAIRQ (MicrosecondsToCycles(3))
+#else
+  #define DELTAIRQ (MicrosecondsToCycles(2))
+#endif
+
+static int startPin = 0;
+static int endPin = 0;
+
 static ICACHE_RAM_ATTR void timer1Interrupt() {
-  uint32_t nextEventCycles;
-  #if F_CPU == 160000000
-  uint8_t cnt = 20;
-  #else
-  uint8_t cnt = 10;
-  #endif
-
-  do {
+  uint32_t nextEventCycles = MicrosecondsToCycles(MAXIRQUS);
+  uint32_t timeoutCycle = GetCycleCountIRQ() + MicrosecondsToCycles(14);
+
+  if (waveformToEnable || waveformToDisable) {
+    // Handle enable/disable requests from main app.
+    waveformEnabled = (waveformEnabled & ~waveformToDisable) | waveformToEnable; // Set the requested waveforms on/off
+    waveformState &= ~waveformToEnable;  // And clear the state of any just started
+    waveformToEnable = 0;
+    waveformToDisable = 0;
+    startPin = __builtin_ffs(waveformEnabled) - 1;
+    endPin = 32 - __builtin_clz(waveformEnabled);
```

Comment:
I don't understand these two lines. Maybe add comments?

Reply:
I'm happily surprised to stump you. :) They're GCC's Find-First-Set (plus one) and Count-Leading-Zeros.  I'll add a comment.